src/iso.c

```diff
@@ -0,0 +1,489 @@
+ * @param iso_p Pointer to the struct iso_point to begin reading at.
+ * @return The number of elements read in, -1 on error.
+ */
+int iso_media_read(void *dest, int elem_size, int num_elem, struct iso_point *stream) {
+    int bytes_needed = num_elem * elem_size;
+    if(bytes_needed <= 0) {
+        return 0;
+    }
+
+    int to_read_from_start_of_cur_extent = stream->cur_offset + bytes_needed;
+    int atapi_blocks_to_read = to_read_from_start_of_cur_extent / ATAPI_BLOCKSIZE;
+
+    if (to_read_from_start_of_cur_extent % ATAPI_BLOCKSIZE) {
+        //integer division rounds down, but we need to round up
+        //if there is rounding to be done
+        atapi_blocks_to_read++;
+    }
+
+    //ensure that we need to perform the read before we do
+    if (atapi_blocks_to_read > 1 ||
+        global_atapi_extent != stream->cur_extent ||
+        global_atapi_unit != stream->ata_unit) {
+        char buffer[atapi_blocks_to_read * ATAPI_BLOCKSIZE];
+        if (!atapi_read(stream->ata_unit, buffer, atapi_blocks_to_read, stream->cur_extent)) {
+            console_printf("atapi_read is 0 in iso_media_read()\n");
+            return -1;
+        }
+        else {
+            //Do not start at the start of the buffer, because that is the start of
+            //the current extent, not where the cur_extent + cur_offset is
+            memcpy(dest, buffer + stream->cur_offset, elem_size * num_elem);
+
+            //copy the last block read in into the global "cache"
+            memcpy(global_atapi_block, buffer + (ATAPI_BLOCKSIZE * (atapi_blocks_to_read - 1)), ATAPI_BLOCKSIZE);
+
+            //update stream, as iso_seek is a mock call to keep track of
+            //"where we are" on the ISO image and does not actually move
+            //the the CD-reading head to read the ISO image, atapi_read does that
+            iso_media_seek(stream, bytes_needed, SEEK_CUR);
+            global_atapi_unit = stream->ata_unit;
+            global_atapi_extent = stream->cur_extent;
+            return num_elem;
+        }
+    }
+    else {
+        //Do not start at the start of the buffer, because that is the start of
+        //the current extent, not where the cur_extent + cur_offset is
+        memcpy(dest, global_atapi_block + stream->cur_offset, elem_size * num_elem);
+
+        //update stream, as iso_seek is a mock call to keep track of
+        //"where we are" on the ISO image and does not actually move
+        //the the CD-reading head to read the ISO image, atapi_read does that
+        iso_media_seek(stream, bytes_needed, SEEK_CUR);
+        return num_elem;
+    }
+}
+
+/**
+ * @brief Moves the iso_p
+ * @details Changes the offset and extent values of the iso_point as instructed
+ *
+ * @param iso_p Pointer to the iso_point to adjust
+ * @param offset How far to move the current pointer
+ * @param whence The relative place to move the pointer from, SEEK_CUR is moving forward from current position, SEEK_SET is from the zeroth byte of the iso
```

Comment:
np: naming? Seems like mode would be better, as thats what you call it in the error below.